lib/Sema/CSSimplify.cpp

```diff
@@ -3449,24 +3448,95 @@ ConstraintSystem::simplifyMemberConstraint(ConstraintKind kind,
     instanceTy = MTT->getInstanceType();
   
   // Value member lookup has some hacks too.
-  if (shouldAttemptFixes() && baseObjTy->getOptionalObjectType()) {
+  auto unwrappedBaseObjTy = baseObjTy->getOptionalObjectType();
+  if (unwrappedBaseObjTy) {
     // If the base type was an optional, look through it.
-    
-    // Determine whether or not we want to provide an optional chaining fixit or
-    // a force unwrap fixit.
-    bool optionalChain;
-    if (!getContextualType())
-      optionalChain = !(Options & ConstraintSystemFlags::PreferForceUnwrapToOptional);
-    else
-      optionalChain = !getContextualType()->getOptionalObjectType().isNull();
-    auto fixKind = optionalChain ? FixKind::OptionalChaining : FixKind::ForceOptional;
-
-    // Note the fix.
-    if (recordFix(fixKind, locator))
-      return SolutionKind::Error;
-    
+
+    // Figure out if we have an existing optional chaining fix done earlier in
+    // this dotted expression chain.
+    bool continuingChain = false;
+    if (auto dotExpr = dyn_cast<UnresolvedDotExpr>(locator->getAnchor())) {
+      for (auto pair : Fixes) {
+        if (pair.first.getKind() == FixKind::OptionalChaining) {
```

Comment:
CSSimplify is too early to be examining the Fixes generated by a particular constraint system.  This should be done in CSApply.

Reply:
This is to avoid creating disjunctions at every step in the chain, which seems worthwhile to avoid 2^n attempted solves. It's true that if we wanted to let that happen we could add a fix at each dot expr and then check fixes in CSApply to make all but the first one just produce no fixit.

Reply:
My thinking is that potentially doing n^2~ish work on every simplification isn't that great either.  If the constraint system doesn't look right, perhaps we should look into changing its morphology or emitting a "compound fix" like you mention here.

Besides, I suspect `n` would be quite low in practice and would be helped by deforestation of the solution space in other parts of this file (This Is Not An Endorsement™).

Reply:
Wait. Sorry, it's late. Just 2n solves because only the optional branch divides again each time. :) 

But the other issue I just remembered here is that ```shouldAttemptFixes()``` is only ever true at the top level, examining disjunctions never wants to allow them, which is the other reason for doing it this way - so that there is only a single initial fix that starts the chain.

Reply:
Now that I've slept, let me try to talk about this better:

Given an example like takeOpt(u.t.s.value) the parameter takes an optional so we want the result to be a fixit to optional chain this series of dots. Stepwise for each dot expr we see an optional base object type, fix it, and tell the constraint system the result should then be optional.

If this was done in a symmetrical, consistent way instead of the admittedly hacky look-to-see-if-a-chain-fix-already-exists way, then it'd have two major impacts. First, we'd need to allow fixes during evaluation of disjunction candidates instead of just at the top level which is an algorithmic change I don't know enough to judge. This may or may not be a good idea in general?

Second, it always type checks in this sort of expression to force the optional at any of the steps and then rewrap at the end, and with multiple independent chaining fixes, a solution of a single force fix at the beginning will get the best score. So taking a symmetrical approach would require additional state on the Disjunction or perhaps on each Constraint inside that would be examined by the scorer to make it so that a forcing fix was scored as some multiple N-times worse than a step of chaining fix. If this ability to make lopsided disjunction scoring seems like it'd be useful elsewhere, maybe that's the way to go?